libraries/chain/fork_database.cpp

```diff
@@ -571,28 +679,75 @@ namespace eosio::chain {
                   "block state not in fork database; cannot mark as valid",
                   ("id", h->id()) );
 
-      by_id_idx.modify( itr, []( bsp& _bsp ) {
-         _bsp->set_valid(true);
+      by_id_idx.modify( itr, []( auto& i ) {
+         BSAccessor::set_valid(*i, true);
       } );
 
-      auto candidate = index.template get<by_lib_block_num>().begin();
+      auto candidate = index.template get<by_best_branch>().begin();
       if( first_preferred( **candidate, *head ) ) {
          head = *candidate;
       }
    }
 
-   template<class bsp>
-   bsp fork_database_t<bsp>::get_block(const block_id_type& id) const {
+   template<class BSP>
+   void fork_database_t<BSP>::update_best_qc( const block_id_type& id, const qc_claim_t& most_recent_ancestor_with_qc ) {
+      std::lock_guard g( my->mtx );
+      my->update_best_qc_impl( id, most_recent_ancestor_with_qc );
+   }
+
+   template<class BSP>
+   void fork_database_impl<BSP>::update_best_qc_impl( const block_id_type& id, const qc_claim_t& most_recent_ancestor_with_qc ) {
+      auto& by_id_idx = index.template get<by_block_id>();
+
+      auto itr = by_id_idx.find( id );
+      EOS_ASSERT( itr != by_id_idx.end(), fork_database_exception,
+                  "block state not in fork database; cannot update", ("id", id) );
+
+      if (BSAccessor::qc_claim_update_needed(**itr, most_recent_ancestor_with_qc)) {
+         by_id_idx.modify( itr, [&]( auto& i ) {
+            BSAccessor::update_best_qc(*i, most_recent_ancestor_with_qc);
+         } );
+      }
+
+      // process descendants
+      vector<block_id_type> descendants;
+      descendants.reserve(index.size());
+      descendants.push_back(id);
+      auto& previdx = index.template get<by_prev>();
+      for( uint32_t i = 0; i < descendants.size(); ++i ) {
+         auto previtr = previdx.lower_bound( descendants[i] );
+         while( previtr != previdx.end() && (*previtr)->previous() == descendants[i] ) {
+            if (BSAccessor::qc_claim_update_needed(**previtr, most_recent_ancestor_with_qc)) {
+               previdx.modify( previtr, [&](auto& i) {
+                  BSAccessor::update_best_qc(*i, most_recent_ancestor_with_qc);
+               });
+            } else {
+               break;
+            }
+
+            descendants.emplace_back( (*previtr)->id() );
+            ++previtr;
+         }
+      }
```

Comment:
I assume this is to avoid a recursive implementation, but imo the added complexity and difficulty to ensure correctness makes me think a recursive solution would be better.

Reply:
This implementation was chosen to match existing implementation in `remove_impl`. We know `remove_impl` works.